File: exercises/static/exercises/montecarlo_visual_loc/python_template/ros2_humble/map.py
import numpy as np
import math
from math import pi as pi
import cv2

class Map:
	def __init__(self, pose3d):
		self.pose3d = pose3d
	
	def RTx(self, angle, tx, ty, tz):
		RT = np.matrix([[1, 0, 0, tx], [0, math.cos(angle), -math.sin(angle), ty], 
						[0, math.sin(angle), math.cos(angle), tz], [0, 0, 0, 1]])
		return RT
		
	def RTy(self, angle, tx, ty, tz):
		RT = np.matrix([[math.cos(angle), 0, math.sin(angle), tx], [0, 1, 0, ty],
						[-math.sin(angle), 0, math.cos(angle), tz], [0, 0, 0, 1]])
		return RT
		
	def RTz(self, angle, tx, ty, tz):
		RT = np.matrix([[math.cos(angle), -math.sin(angle), 0, tx], [math.sin(angle), math.cos(angle), 0, ty],
						[0, 0, 1, tz], [0, 0, 0, 1]])
		return RT
		
	def RTVacuum(self):
		RTz = self.RTz(pi/2, 50, 70, 0)
		return RTz
		
	def getRobotCoordinates(self):
		pose = self.pose3d()
		x = pose.x
		y = pose.y
		
		scale_y = 15; offset_y = 63
		y = scale_y * y + offset_y
		
		scale_x = -30; offset_x = 171
		x = scale_x * x + offset_x
		
		return x, y

	def getRobotAngle(self):
		pose = self.pose3d()

		return pose.yaw,

	# Function to reset
	def reset(self):
		# Nothing to do, service takes care!
		pass
	

File: exercises/montecarlo_visual_loc/cpp_lib/src/Map.cpp
#include "Map.hpp"
#include <cmath>

Map::Map(std::function<Pose3d()> pose3d)
    : pose3d_(pose3d)
{
}

cv::Mat Map::RTx(double angle, double tx, double ty, double tz)
{
    return (cv::Mat_<double>(4, 4) << 
        1.0, 0.0, 0.0, tx,
        0.0, std::cos(angle), -std::sin(angle), ty,
        0.0, std::sin(angle), std::cos(angle), tz,
        0.0, 0.0, 0.0, 1.0);
}

cv::Mat Map::RTy(double angle, double tx, double ty, double tz)
{
    return (cv::Mat_<double>(4, 4) << 
        std::cos(angle), 0.0, std::sin(angle), tx,
        0.0, 1.0, 0.0, ty,
        -std::sin(angle), 0.0, std::cos(angle), tz,
        0.0, 0.0, 0.0, 1.0);
}

cv::Mat Map::RTz(double angle, double tx, double ty, double tz)
{
    return (cv::Mat_<double>(4, 4) << 
        std::cos(angle), -std::sin(angle), 0.0, tx,
        std::sin(angle), std::cos(angle), 0.0, ty,
        0.0, 0.0, 1.0, tz,
        0.0, 0.0, 0.0, 1.0);
}

cv::Mat Map::RTVacuum()
{
    return RTz(CV_PI / 2.0, 50.0, 70.0, 0.0);
}

std::array<double, 2> Map::getRobotCoordinates()
{
    Pose3d pose = pose3d_();
    double x = pose.x;
    double y = pose.y;

    double scale_y = 15.0;
    double offset_y = 63.0;
    y = scale_y * y + offset_y;

    double scale_x = -30.0;
    double offset_x = 171.0;
    x = scale_x * x + offset_x;

    return {x, y};
}

double Map::getRobotAngle()
{
    Pose3d pose = pose3d_();
    return pose.yaw;
}

// Function to reset
void Map::reset()
{
    // Nothing to do, service takes care!
}